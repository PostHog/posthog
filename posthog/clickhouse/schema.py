# This file contains all CREATE TABLE queries, used to sync and test schema
import re

from posthog.clickhouse.dead_letter_queue import *
from posthog.clickhouse.plugin_log_entries import *
from posthog.models.cohort.sql import *
from posthog.models.event.sql import *
from posthog.models.group.sql import *
from posthog.models.person.sql import *
from posthog.models.session_recording_event.sql import *

CREATE_MERGETREE_TABLE_QUERIES = (
    CREATE_COHORTPEOPLE_TABLE_SQL,
    PERSON_STATIC_COHORT_TABLE_SQL,
    DEAD_LETTER_QUEUE_TABLE_SQL,
    EVENTS_TABLE_SQL,
    GROUPS_TABLE_SQL,
    PERSONS_TABLE_SQL,
    PERSONS_DISTINCT_ID_TABLE_SQL,
    PERSON_DISTINCT_ID2_TABLE_SQL,
    PLUGIN_LOG_ENTRIES_TABLE_SQL,
    SESSION_RECORDING_EVENTS_TABLE_SQL,
)
CREATE_DISTRIBUTED_TABLE_QUERIES = (
    WRITABLE_EVENTS_TABLE_SQL,
    DISTRIBUTED_EVENTS_TABLE_SQL,
    WRITABLE_SESSION_RECORDING_EVENTS_TABLE_SQL,
    DISTRIBUTED_SESSION_RECORDING_EVENTS_TABLE_SQL,
)
CREATE_KAFKA_TABLE_QUERIES = (
    KAFKA_DEAD_LETTER_QUEUE_TABLE_SQL,
    KAFKA_EVENTS_TABLE_JSON_SQL,
    KAFKA_GROUPS_TABLE_SQL,
    KAFKA_PERSONS_TABLE_SQL,
    KAFKA_PERSONS_DISTINCT_ID_TABLE_SQL,
    KAFKA_PERSON_DISTINCT_ID2_TABLE_SQL,
    KAFKA_PLUGIN_LOG_ENTRIES_TABLE_SQL,
    KAFKA_SESSION_RECORDING_EVENTS_TABLE_SQL,
)
CREATE_MV_TABLE_QUERIES = (
    DEAD_LETTER_QUEUE_TABLE_MV_SQL,
    EVENTS_TABLE_JSON_MV_SQL,
    GROUPS_TABLE_MV_SQL,
    PERSONS_TABLE_MV_SQL,
    PERSONS_DISTINCT_ID_TABLE_MV_SQL,
    PERSON_DISTINCT_ID2_MV_SQL,
    PLUGIN_LOG_ENTRIES_TABLE_MV_SQL,
    SESSION_RECORDING_EVENTS_TABLE_MV_SQL,
)

CREATE_TABLE_QUERIES = (
    CREATE_MERGETREE_TABLE_QUERIES
    + CREATE_DISTRIBUTED_TABLE_QUERIES
    + CREATE_KAFKA_TABLE_QUERIES
    + CREATE_MV_TABLE_QUERIES
)

build_query = lambda query: query if isinstance(query, str) else query()
get_table_name = lambda query: re.findall(r" ([a-z0-9_]+) ON CLUSTER", build_query(query))[0]
