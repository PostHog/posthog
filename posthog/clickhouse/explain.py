import json
from dataclasses import dataclass

from posthog.clickhouse.client import sync_execute
from posthog.hogql.context import HogQLContext
from posthog.schema import QueryIndexUsage


def find_all_reads(explain: dict) -> list[dict]:
    """
    Looks for a Plan/Subplan with "Indexes" element. ClickHouse' MergeTree engine uses "ReadFromMergeTree"
    plan Node Type to describe reading from a table and if defined with index it will contain "Indexes" field.
    """
    reads = []
    plan = explain.get("Plan", explain)
    if "Indexes" in plan:
        reads.append(plan)
    for subplan in plan.get("Plans", []):
        reads += find_all_reads(subplan)
    return reads


def selected_less_granules(index: dict, tiny_data_granules: int = 100) -> bool:
    """
    Each "Indexes" field contains a description of how the index impact selection of partitions and granules.
    Index is effective if the number of granules selected is smaller than before using an index. If the data is
    small enough, the index won't limit what we read.
    """
    initial_granules = index.get("Initial Granules", 0)
    return index.get("Selected Granules", 0) < initial_granules or initial_granules < tiny_data_granules


@dataclass
class ReadIndexUsage:
    table: str
    use: QueryIndexUsage


def guestimate_index_use(plan_with_indexes: dict) -> ReadIndexUsage:
    """
    For a given table read we try to indentify if an index was used. This is limited as the plan is being processed
    without a context (a table schema / index). Some tables have simple index and no real partitioning.
    :param plan_with_indexes:
    :return:
    """
    db_table = plan_with_indexes.get("Description", "")
    result = ReadIndexUsage(table=db_table, use=QueryIndexUsage.NO)
    if "Indexes" not in plan_with_indexes:
        return result

    indexes = plan_with_indexes.get("Indexes", [])

    if db_table.endswith(".person_distinct_id_overrides"):
        if len(indexes) == 1:
            index = indexes[0]
            if (
                index.get("Condition", "") != "true"
                and "team_id" in index.get("Keys", [])
                and selected_less_granules(index)
            ):
                result.use = QueryIndexUsage.YES

        return result
    elif db_table.endswith(".sharded_events"):
        min_max = False
        partition = False
        primary_key = False
        for index in indexes:
            if index.get("Condition", "") == "true":  # if the condition for index was not set
                continue
            index_type = index.get("Type", "")
            if index_type == "MinMax":
                min_max = selected_less_granules(index)
            elif index_type == "Partition":
                partition = selected_less_granules(index)
            elif index_type == "PrimaryKey":
                primary_key = len(index.get("Keys", [])) > 0 and selected_less_granules(index)
        if (min_max or partition) and primary_key:
            result.use = QueryIndexUsage.YES

        return result

    result.use = QueryIndexUsage.UNDECISIVE
    has_min_max = False
    min_max = False
    has_partition = False
    partition = False
    primary_key = False
    for index in indexes:
        if index.get("Condition", "") == "true":
            continue
        index_type = index.get("Type", "")
        if index_type == "MinMax":
            has_min_max = True
            min_max = selected_less_granules(index)
        elif index_type == "Partition":
            has_partition = True
            partition = selected_less_granules(index)
        elif index_type == "PrimaryKey":
            primary_key = len(index.get("Keys", [])) > 0 and selected_less_granules(index)
    if primary_key:
        if (not has_min_max and not has_partition) or min_max or partition:
            result.use = QueryIndexUsage.YES
        else:
            result.use = QueryIndexUsage.PARTIAL

    return result


def extract_index_usage_from_plan(plan: str) -> QueryIndexUsage:
    try:
        explain = json.loads(plan)
        all_indices_use = [guestimate_index_use(r) for r in find_all_reads(explain[0])]
        if all(x.use == QueryIndexUsage.YES for x in all_indices_use):
            return QueryIndexUsage.YES
        elif any(x.use == QueryIndexUsage.YES for x in all_indices_use):
            return QueryIndexUsage.PARTIAL
        elif all(x.use == QueryIndexUsage.NO for x in all_indices_use):
            return QueryIndexUsage.NO
    except json.decoder.JSONDecodeError:
        pass

    return QueryIndexUsage.UNDECISIVE


def execute_explain_get_index_use(clickhouse_sql: str, context: HogQLContext) -> QueryIndexUsage:
    # try:
    explain_results = sync_execute(
        f"EXPLAIN PLAN indexes=1,json=1 {clickhouse_sql}",
        context.values,
        with_column_types=True,
        team_id=context.team_id,
        readonly=True,
    )
    return extract_index_usage_from_plan(explain_results[0][0][0])


# except:
#     return QueryIndexUsage.UNDECISIVE
