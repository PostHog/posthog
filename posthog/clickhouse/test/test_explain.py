import json
from dataclasses import dataclass

import pytest

from posthog.clickhouse.explain import (
    find_all_reads,
    guestimate_index_use,
    extract_index_usage_from_plan,
    ReadIndexUsage,
)
from posthog.schema import QueryIndexUsage


@dataclass
class TestCaseData:
    query: str
    reads: int
    reads_use: list[ReadIndexUsage]
    use: QueryIndexUsage
    plan: str


test_cases: list[TestCaseData] = [
    TestCaseData(
        query="a proper query with all filters on events",
        reads=1,
        reads_use=[ReadIndexUsage(table="posthog.sharded_events", use=QueryIndexUsage.YES)],
        use=QueryIndexUsage.YES,
        plan="""[
  {
    "Plan": {
      "Node Type": "Union",
      "Plans": [
        {
          "Node Type": "Expression",
          "Description": "(Projection + Before ORDER BY)",
          "Plans": [
            {
              "Node Type": "Expression",
              "Plans": [
                {
                  "Node Type": "ReadFromMergeTree",
                  "Description": "posthog.sharded_events",
                  "Indexes": [
                    {
                      "Type": "MinMax",
                      "Keys": [
                        "timestamp"
                      ],
                      "Condition": "(timestamp in ('1735828351', +Inf))",
                      "Initial Parts": 2799,
                      "Selected Parts": 222,
                      "Initial Granules": 173418552,
                      "Selected Granules": 26462626
                    },
                    {
                      "Type": "Partition",
                      "Keys": [
                        "toYYYYMM(timestamp)"
                      ],
                      "Condition": "(toYYYYMM(timestamp) in [202501, +Inf))",
                      "Initial Parts": 222,
                      "Selected Parts": 222,
                      "Initial Granules": 26462626,
                      "Selected Granules": 26462626
                    },
                    {
                      "Type": "PrimaryKey",
                      "Keys": [
                        "team_id",
                        "toDate(timestamp)",
                        "event"
                      ],
                      "Condition": "and((toDate(timestamp) in [20090, +Inf)), and((event in ['$pageview', '$pageview']), (team_id in [2, 2])))",
                      "Initial Parts": 222,
                      "Selected Parts": 181,
                      "Initial Granules": 26462626,
                      "Selected Granules": 35950
                    }
                  ]
                }
              ]
            }
          ]
        },
        {
          "Node Type": "ReadFromRemote",
          "Description": "Read from remote replica"
        }
      ]
    }
  }
]
""",
    ),
    TestCaseData(
        query="missing timestamp condition",
        reads=1,
        reads_use=[ReadIndexUsage(table="posthog.sharded_events", use=QueryIndexUsage.NO)],
        use=QueryIndexUsage.NO,
        plan="""
[
  {
    "Plan": {
      "Node Type": "Union",
      "Plans": [
        {
          "Node Type": "Expression",
          "Description": "(Projection + Before ORDER BY)",
          "Plans": [
            {
              "Node Type": "Expression",
              "Plans": [
                {
                  "Node Type": "ReadFromMergeTree",
                  "Description": "posthog.sharded_events",
                  "Indexes": [
                    {
                      "Type": "MinMax",
                      "Condition": "true",
                      "Initial Parts": 2972,
                      "Selected Parts": 2972,
                      "Initial Granules": 171921031,
                      "Selected Granules": 171921031
                    },
                    {
                      "Type": "Partition",
                      "Condition": "true",
                      "Initial Parts": 2972,
                      "Selected Parts": 2972,
                      "Initial Granules": 171921031,
                      "Selected Granules": 171921031
                    },
                    {
                      "Type": "PrimaryKey",
                      "Keys": [
                        "team_id",
                        "event"
                      ],
                      "Condition": "and((event in ['$pageview', '$pageview']), (team_id in [2, 2]))",
                      "Initial Parts": 2972,
                      "Selected Parts": 1002,
                      "Initial Granules": 171921031,
                      "Selected Granules": 245218
                    }
                  ]
                }
              ]
            }
          ]
        },
        {
          "Node Type": "ReadFromRemote",
          "Description": "Read from remote replica"
        }
      ]
    }
  }
]
""",
    ),
    TestCaseData(
        query="production query often failing cause ClickHouse not using index",
        reads=2,
        reads_use=[
            ReadIndexUsage(table="posthog.sharded_events", use=QueryIndexUsage.NO),
            ReadIndexUsage(table="posthog.person_distinct_id_overrides", use=QueryIndexUsage.YES),
        ],
        use=QueryIndexUsage.PARTIAL,
        plan="""
[
  {
    "Plan": {
      "Node Type": "Expression",
      "Description": "Projection",
      "Plans": [
        {
          "Node Type": "Limit",
          "Description": "preliminary LIMIT (without OFFSET)",
          "Plans": [
            {
              "Node Type": "Sorting",
              "Description": "Sorting for ORDER BY",
              "Plans": [
                {
                  "Node Type": "Expression",
                  "Description": "Before ORDER BY",
                  "Plans": [
                    {
                      "Node Type": "Aggregating",
                      "Plans": [
                        {
                          "Node Type": "Expression",
                          "Description": "(Before GROUP BY + (Projection + Before ORDER BY))",
                          "Plans": [
                            {
                              "Node Type": "MergingAggregated",
                              "Plans": [
                                {
                                  "Node Type": "Union",
                                  "Plans": [
                                    {
                                      "Node Type": "Aggregating",
                                      "Plans": [
                                        {
                                          "Node Type": "Expression",
                                          "Description": "(Before GROUP BY + )",
                                          "Plans": [
                                            {
                                              "Node Type": "Join",
                                              "Description": "JOIN FillRightFirst",
                                              "Plans": [
                                                {
                                                  "Node Type": "Expression",
                                                  "Plans": [
                                                    {
                                                      "Node Type": "ReadFromMergeTree",
                                                      "Description": "posthog.sharded_events",
                                                      "Indexes": [
                                                        {
                                                          "Type": "MinMax",
                                                          "Condition": "true",
                                                          "Initial Parts": 2759,
                                                          "Selected Parts": 2759,
                                                          "Initial Granules": 173438060,
                                                          "Selected Granules": 173438060
                                                        },
                                                        {
                                                          "Type": "Partition",
                                                          "Condition": "true",
                                                          "Initial Parts": 2759,
                                                          "Selected Parts": 2759,
                                                          "Initial Granules": 173438060,
                                                          "Selected Granules": 173438060
                                                        },
                                                        {
                                                          "Type": "PrimaryKey",
                                                          "Keys": [
                                                            "team_id"
                                                          ],
                                                          "Condition": "(team_id in [16617, 16617])",
                                                          "Initial Parts": 2759,
                                                          "Selected Parts": 1699,
                                                          "Initial Granules": 173438060,
                                                          "Selected Granules": 1877152
                                                        }
                                                      ]
                                                    }
                                                  ]
                                                },
                                                {
                                                  "Node Type": "Expression",
                                                  "Description": "(Joined actions + (Rename joined columns + (Projection + Before ORDER BY)))",
                                                  "Plans": [
                                                    {
                                                      "Node Type": "Filter",
                                                      "Description": "HAVING",
                                                      "Plans": [
                                                        {
                                                          "Node Type": "Aggregating",
                                                          "Plans": [
                                                            {
                                                              "Node Type": "Expression",
                                                              "Description": "Before GROUP BY",
                                                              "Plans": [
                                                                {
                                                                  "Node Type": "Expression",
                                                                  "Plans": [
                                                                    {
                                                                      "Node Type": "ReadFromMergeTree",
                                                                      "Description": "posthog.person_distinct_id_overrides",
                                                                      "Indexes": [
                                                                        {
                                                                          "Type": "PrimaryKey",
                                                                          "Keys": [
                                                                            "team_id"
                                                                          ],
                                                                          "Condition": "(team_id in [16617, 16617])",
                                                                          "Initial Parts": 18,
                                                                          "Selected Parts": 18,
                                                                          "Initial Granules": 155239,
                                                                          "Selected Granules": 18
                                                                        }
                                                                      ]
                                                                    }
                                                                  ]
                                                                }
                                                              ]
                                                            }
                                                          ]
                                                        }
                                                      ]
                                                    }
                                                  ]
                                                }
                                              ]
                                            }
                                          ]
                                        }
                                      ]
                                    },
                                    {
                                      "Node Type": "ReadFromRemote",
                                      "Description": "Read from remote replica"
                                    }
                                  ]
                                }
                              ]
                            }
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  }
]
""",
    ),
    TestCaseData(
        query="local query little granules",
        reads=1,
        reads_use=[ReadIndexUsage(table="default.sharded_events", use=QueryIndexUsage.NO)],
        use=QueryIndexUsage.NO,
        plan="""
[
  {
    "Plan": {
      "Node Type": "Expression",
      "Description": "(Project names + Projection)",
      "Plans": [
        {
          "Node Type": "Limit",
          "Description": "preliminary LIMIT (without OFFSET)",
          "Plans": [
            {
              "Node Type": "Expression",
              "Plans": [
                {
                  "Node Type": "ReadFromMergeTree",
                  "Description": "default.sharded_events",
                  "Indexes": [
                    {
                      "Type": "MinMax",
                      "Condition": "true",
                      "Initial Parts": 12,
                      "Selected Parts": 12,
                      "Initial Granules": 39,
                      "Selected Granules": 39
                    },
                    {
                      "Type": "Partition",
                      "Condition": "true",
                      "Initial Parts": 12,
                      "Selected Parts": 12,
                      "Initial Granules": 39,
                      "Selected Granules": 39
                    },
                    {
                      "Type": "PrimaryKey",
                      "Keys": ["team_id"],
                      "Condition": "(team_id in [1, 1])",
                      "Initial Parts": 12,
                      "Selected Parts": 9,
                      "Initial Granules": 39,
                      "Selected Granules": 32
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  }
]
""",
    ),
    TestCaseData(
        query="local query little granules, using index",
        reads=1,
        reads_use=[ReadIndexUsage(table="default.sharded_events", use=QueryIndexUsage.YES)],
        use=QueryIndexUsage.YES,
        plan="""
[
  {
    "Plan": {
      "Node Type": "Expression",
      "Description": "(Project names + Projection)",
      "Plans": [
        {
          "Node Type": "Limit",
          "Description": "preliminary LIMIT (without OFFSET)",
          "Plans": [
            {
              "Node Type": "Expression",
              "Plans": [
                {
                  "Node Type": "ReadFromMergeTree",
                  "Description": "default.sharded_events",
                  "Indexes": [
                    {
                      "Type": "MinMax",
                      "Keys": ["timestamp"],
                      "Condition": "(toTimezone(timestamp, 'UTC') in ('1735831217', +Inf))",
                      "Initial Parts": 11,
                      "Selected Parts": 6,
                      "Initial Granules": 38,
                      "Selected Granules": 32
                    },
                    {
                      "Type": "Partition",
                      "Condition": "true",
                      "Initial Parts": 6,
                      "Selected Parts": 6,
                      "Initial Granules": 32,
                      "Selected Granules": 32
                    },
                    {
                      "Type": "PrimaryKey",
                      "Keys": ["team_id"],
                      "Condition": "(team_id in [1, 1])",
                      "Initial Parts": 6,
                      "Selected Parts": 5,
                      "Initial Granules": 32,
                      "Selected Granules": 27
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  }
]
""",
    ),
]


@pytest.mark.parametrize("case", test_cases, ids=[x.query for x in test_cases])
def test_full_queries(case):
    explain = json.loads(case.plan)
    reads = find_all_reads(explain[0])
    assert case.reads == len(reads)
    assert case.reads_use == [guestimate_index_use(r) for r in reads]
    assert case.use == extract_index_usage_from_plan(case.plan)
