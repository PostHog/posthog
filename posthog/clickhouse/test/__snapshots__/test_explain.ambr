# serializer version: 1
# name: test_full_queries[a proper query with all filters on events]
  '''
  [
    {
      "Plan": {
        "Node Type": "Union",
        "Plans": [
          {
            "Node Type": "Expression",
            "Description": "(Projection + Before ORDER BY)",
            "Plans": [
              {
                "Node Type": "Expression",
                "Plans": [
                  {
                    "Node Type": "ReadFromMergeTree",
                    "Description": "posthog.sharded_events",
                    "Indexes": [
                      {
                        "Type": "MinMax",
                        "Keys": [
                          "timestamp"
                        ],
                        "Condition": "(timestamp in ('1735828351', +Inf))",
                        "Initial Parts": 2799,
                        "Selected Parts": 222,
                        "Initial Granules": 173418552,
                        "Selected Granules": 26462626
                      },
                      {
                        "Type": "Partition",
                        "Keys": [
                          "toYYYYMM(timestamp)"
                        ],
                        "Condition": "(toYYYYMM(timestamp) in [202501, +Inf))",
                        "Initial Parts": 222,
                        "Selected Parts": 222,
                        "Initial Granules": 26462626,
                        "Selected Granules": 26462626
                      },
                      {
                        "Type": "PrimaryKey",
                        "Keys": [
                          "team_id",
                          "toDate(timestamp)",
                          "event"
                        ],
                        "Condition": "and((toDate(timestamp) in [20090, +Inf)), and((event in ['$pageview', '$pageview']), (team_id in [2, 2])))",
                        "Initial Parts": 222,
                        "Selected Parts": 181,
                        "Initial Granules": 26462626,
                        "Selected Granules": 35950
                      }
                    ]
                  }
                ]
              }
            ]
          },
          {
            "Node Type": "ReadFromRemote",
            "Description": "Read from remote replica"
          }
        ]
      }
    }
  ]
  
  '''
# ---
# name: test_full_queries[a proper query with all filters on events].1
  1
# ---
# name: test_full_queries[a proper query with all filters on events].2
  list([
    ReadIndexUsage(table='posthog.sharded_events', use=<QueryIndexUsage.YES: 'yes'>),
  ])
# ---
# name: test_full_queries[a proper query with all filters on events].3
  <QueryIndexUsage.YES: 'yes'>
# ---
# name: test_full_queries[local query little granules, using index]
  '''
  
  [
    {
      "Plan": {
        "Node Type": "Expression",
        "Description": "(Project names + Projection)",
        "Plans": [
          {
            "Node Type": "Limit",
            "Description": "preliminary LIMIT (without OFFSET)",
            "Plans": [
              {
                "Node Type": "Expression",
                "Plans": [
                  {
                    "Node Type": "ReadFromMergeTree",
                    "Description": "default.sharded_events",
                    "Indexes": [
                      {
                        "Type": "MinMax",
                        "Keys": ["timestamp"],
                        "Condition": "(toTimezone(timestamp, 'UTC') in ('1735831217', +Inf))",
                        "Initial Parts": 11,
                        "Selected Parts": 6,
                        "Initial Granules": 38,
                        "Selected Granules": 32
                      },
                      {
                        "Type": "Partition",
                        "Condition": "true",
                        "Initial Parts": 6,
                        "Selected Parts": 6,
                        "Initial Granules": 32,
                        "Selected Granules": 32
                      },
                      {
                        "Type": "PrimaryKey",
                        "Keys": ["team_id"],
                        "Condition": "(team_id in [1, 1])",
                        "Initial Parts": 6,
                        "Selected Parts": 5,
                        "Initial Granules": 32,
                        "Selected Granules": 27
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      }
    }
  ]
  
  '''
# ---
# name: test_full_queries[local query little granules, using index].1
  1
# ---
# name: test_full_queries[local query little granules, using index].2
  list([
    ReadIndexUsage(table='default.sharded_events', use=<QueryIndexUsage.YES: 'yes'>),
  ])
# ---
# name: test_full_queries[local query little granules, using index].3
  <QueryIndexUsage.YES: 'yes'>
# ---
# name: test_full_queries[local query little granules]
  '''
  
  [
    {
      "Plan": {
        "Node Type": "Expression",
        "Description": "(Project names + Projection)",
        "Plans": [
          {
            "Node Type": "Limit",
            "Description": "preliminary LIMIT (without OFFSET)",
            "Plans": [
              {
                "Node Type": "Expression",
                "Plans": [
                  {
                    "Node Type": "ReadFromMergeTree",
                    "Description": "default.sharded_events",
                    "Indexes": [
                      {
                        "Type": "MinMax",
                        "Condition": "true",
                        "Initial Parts": 12,
                        "Selected Parts": 12,
                        "Initial Granules": 39,
                        "Selected Granules": 39
                      },
                      {
                        "Type": "Partition",
                        "Condition": "true",
                        "Initial Parts": 12,
                        "Selected Parts": 12,
                        "Initial Granules": 39,
                        "Selected Granules": 39
                      },
                      {
                        "Type": "PrimaryKey",
                        "Keys": ["team_id"],
                        "Condition": "(team_id in [1, 1])",
                        "Initial Parts": 12,
                        "Selected Parts": 9,
                        "Initial Granules": 39,
                        "Selected Granules": 32
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      }
    }
  ]
  
  '''
# ---
# name: test_full_queries[local query little granules].1
  1
# ---
# name: test_full_queries[local query little granules].2
  list([
    ReadIndexUsage(table='default.sharded_events', use=<QueryIndexUsage.NO: 'no'>),
  ])
# ---
# name: test_full_queries[local query little granules].3
  <QueryIndexUsage.NO: 'no'>
# ---
# name: test_full_queries[missing timestamp condition]
  '''
  
  [
    {
      "Plan": {
        "Node Type": "Union",
        "Plans": [
          {
            "Node Type": "Expression",
            "Description": "(Projection + Before ORDER BY)",
            "Plans": [
              {
                "Node Type": "Expression",
                "Plans": [
                  {
                    "Node Type": "ReadFromMergeTree",
                    "Description": "posthog.sharded_events",
                    "Indexes": [
                      {
                        "Type": "MinMax",
                        "Condition": "true",
                        "Initial Parts": 2972,
                        "Selected Parts": 2972,
                        "Initial Granules": 171921031,
                        "Selected Granules": 171921031
                      },
                      {
                        "Type": "Partition",
                        "Condition": "true",
                        "Initial Parts": 2972,
                        "Selected Parts": 2972,
                        "Initial Granules": 171921031,
                        "Selected Granules": 171921031
                      },
                      {
                        "Type": "PrimaryKey",
                        "Keys": [
                          "team_id",
                          "event"
                        ],
                        "Condition": "and((event in ['$pageview', '$pageview']), (team_id in [2, 2]))",
                        "Initial Parts": 2972,
                        "Selected Parts": 1002,
                        "Initial Granules": 171921031,
                        "Selected Granules": 245218
                      }
                    ]
                  }
                ]
              }
            ]
          },
          {
            "Node Type": "ReadFromRemote",
            "Description": "Read from remote replica"
          }
        ]
      }
    }
  ]
  
  '''
# ---
# name: test_full_queries[missing timestamp condition].1
  1
# ---
# name: test_full_queries[missing timestamp condition].2
  list([
    ReadIndexUsage(table='posthog.sharded_events', use=<QueryIndexUsage.NO: 'no'>),
  ])
# ---
# name: test_full_queries[missing timestamp condition].3
  <QueryIndexUsage.NO: 'no'>
# ---
# name: test_full_queries[production query often failing cause ClickHouse not using index]
  '''
  
  [
    {
      "Plan": {
        "Node Type": "Expression",
        "Description": "Projection",
        "Plans": [
          {
            "Node Type": "Limit",
            "Description": "preliminary LIMIT (without OFFSET)",
            "Plans": [
              {
                "Node Type": "Sorting",
                "Description": "Sorting for ORDER BY",
                "Plans": [
                  {
                    "Node Type": "Expression",
                    "Description": "Before ORDER BY",
                    "Plans": [
                      {
                        "Node Type": "Aggregating",
                        "Plans": [
                          {
                            "Node Type": "Expression",
                            "Description": "(Before GROUP BY + (Projection + Before ORDER BY))",
                            "Plans": [
                              {
                                "Node Type": "MergingAggregated",
                                "Plans": [
                                  {
                                    "Node Type": "Union",
                                    "Plans": [
                                      {
                                        "Node Type": "Aggregating",
                                        "Plans": [
                                          {
                                            "Node Type": "Expression",
                                            "Description": "(Before GROUP BY + )",
                                            "Plans": [
                                              {
                                                "Node Type": "Join",
                                                "Description": "JOIN FillRightFirst",
                                                "Plans": [
                                                  {
                                                    "Node Type": "Expression",
                                                    "Plans": [
                                                      {
                                                        "Node Type": "ReadFromMergeTree",
                                                        "Description": "posthog.sharded_events",
                                                        "Indexes": [
                                                          {
                                                            "Type": "MinMax",
                                                            "Condition": "true",
                                                            "Initial Parts": 2759,
                                                            "Selected Parts": 2759,
                                                            "Initial Granules": 173438060,
                                                            "Selected Granules": 173438060
                                                          },
                                                          {
                                                            "Type": "Partition",
                                                            "Condition": "true",
                                                            "Initial Parts": 2759,
                                                            "Selected Parts": 2759,
                                                            "Initial Granules": 173438060,
                                                            "Selected Granules": 173438060
                                                          },
                                                          {
                                                            "Type": "PrimaryKey",
                                                            "Keys": [
                                                              "team_id"
                                                            ],
                                                            "Condition": "(team_id in [16617, 16617])",
                                                            "Initial Parts": 2759,
                                                            "Selected Parts": 1699,
                                                            "Initial Granules": 173438060,
                                                            "Selected Granules": 1877152
                                                          }
                                                        ]
                                                      }
                                                    ]
                                                  },
                                                  {
                                                    "Node Type": "Expression",
                                                    "Description": "(Joined actions + (Rename joined columns + (Projection + Before ORDER BY)))",
                                                    "Plans": [
                                                      {
                                                        "Node Type": "Filter",
                                                        "Description": "HAVING",
                                                        "Plans": [
                                                          {
                                                            "Node Type": "Aggregating",
                                                            "Plans": [
                                                              {
                                                                "Node Type": "Expression",
                                                                "Description": "Before GROUP BY",
                                                                "Plans": [
                                                                  {
                                                                    "Node Type": "Expression",
                                                                    "Plans": [
                                                                      {
                                                                        "Node Type": "ReadFromMergeTree",
                                                                        "Description": "posthog.person_distinct_id_overrides",
                                                                        "Indexes": [
                                                                          {
                                                                            "Type": "PrimaryKey",
                                                                            "Keys": [
                                                                              "team_id"
                                                                            ],
                                                                            "Condition": "(team_id in [16617, 16617])",
                                                                            "Initial Parts": 18,
                                                                            "Selected Parts": 18,
                                                                            "Initial Granules": 155239,
                                                                            "Selected Granules": 18
                                                                          }
                                                                        ]
                                                                      }
                                                                    ]
                                                                  }
                                                                ]
                                                              }
                                                            ]
                                                          }
                                                        ]
                                                      }
                                                    ]
                                                  }
                                                ]
                                              }
                                            ]
                                          }
                                        ]
                                      },
                                      {
                                        "Node Type": "ReadFromRemote",
                                        "Description": "Read from remote replica"
                                      }
                                    ]
                                  }
                                ]
                              }
                            ]
                          }
                        ]
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      }
    }
  ]
  
  '''
# ---
# name: test_full_queries[production query often failing cause ClickHouse not using index].1
  2
# ---
# name: test_full_queries[production query often failing cause ClickHouse not using index].2
  list([
    ReadIndexUsage(table='posthog.sharded_events', use=<QueryIndexUsage.NO: 'no'>),
    ReadIndexUsage(table='posthog.person_distinct_id_overrides', use=<QueryIndexUsage.YES: 'yes'>),
  ])
# ---
# name: test_full_queries[production query often failing cause ClickHouse not using index].3
  <QueryIndexUsage.PARTIAL: 'partial'>
# ---
# name: test_full_queries[session query not using indices]
  '''
  
  [
     {
        "Plan" : {
           "Description" : "(Projection + (Before ORDER BY + (WHERE + (Projection + Before ORDER BY))))",
           "Node Type" : "Expression",
           "Plans" : [
              {
                 "Description" : "preliminary LIMIT (without OFFSET)",
                 "Node Type" : "Limit",
                 "Plans" : [
                    {
                       "Description" : "((WHERE + (Projection + Before ORDER BY)))[split]",
                       "Node Type" : "Filter",
                       "Plans" : [
                          {
                             "Description" : "HAVING",
                             "Node Type" : "Filter",
                             "Plans" : [
                                {
                                   "Node Type" : "MergingAggregated",
                                   "Plans" : [
                                      {
                                         "Node Type" : "Union",
                                         "Plans" : [
                                            {
                                               "Node Type" : "Aggregating",
                                               "Plans" : [
                                                  {
                                                     "Description" : "Before GROUP BY",
                                                     "Node Type" : "Expression",
                                                     "Plans" : [
                                                        {
                                                           "Node Type" : "Expression",
                                                           "Plans" : [
                                                              {
                                                                 "Description" : "posthog.sharded_raw_sessions",
                                                                 "Indexes" : [
                                                                    {
                                                                       "Condition" : "true",
                                                                       "Initial Granules" : 4238362,
                                                                       "Initial Parts" : 1516,
                                                                       "Selected Granules" : 4238362,
                                                                       "Selected Parts" : 1516,
                                                                       "Type" : "MinMax"
                                                                    },
                                                                    {
                                                                       "Condition" : "true",
                                                                       "Initial Granules" : 4238362,
                                                                       "Initial Parts" : 1516,
                                                                       "Selected Granules" : 4238362,
                                                                       "Selected Parts" : 1516,
                                                                       "Type" : "Partition"
                                                                    },
                                                                    {
                                                                       "Condition" : "(team_id in [2, 2])",
                                                                       "Initial Granules" : 4238362,
                                                                       "Initial Parts" : 1516,
                                                                       "Keys" : [
                                                                          "team_id"
                                                                       ],
                                                                       "Selected Granules" : 7616,
                                                                       "Selected Parts" : 150,
                                                                       "Type" : "PrimaryKey"
                                                                    }
                                                                 ],
                                                                 "Node Type" : "ReadFromMergeTree"
                                                              }
                                                           ]
                                                        }
                                                     ]
                                                  }
                                               ]
                                            },
                                            {
                                               "Description" : "Read from remote replica",
                                               "Node Type" : "ReadFromRemote"
                                            }
                                         ]
                                      }
                                   ]
                                }
                             ]
                          }
                       ]
                    }
                 ]
              }
           ]
        }
     }
  ]
  
  '''
# ---
# name: test_full_queries[session query not using indices].1
  1
# ---
# name: test_full_queries[session query not using indices].2
  list([
    ReadIndexUsage(table='posthog.sharded_raw_sessions', use=<QueryIndexUsage.PARTIAL: 'partial'>),
  ])
# ---
# name: test_full_queries[session query not using indices].3
  <QueryIndexUsage.PARTIAL: 'partial'>
# ---
